lib/src/license_detection/new_tokenizer.dart

```diff
@@ -0,0 +1,278 @@
+import 'dart:collection';
+
+import 'package:meta/meta.dart';
+import 'package:source_span/source_span.dart';
+import 'package:string_scanner/string_scanner.dart';
+
+@sealed
+class NewToken {
+  final String normalizedText;
+  final int tokenID;
+  final int line;
+  final FileSpan fileSpan;
+
+  NewToken(this.normalizedText, this.tokenID, this.line, this.fileSpan);
+}
+
+List<NewToken> newTokenizer(String text) {
+  // Guideline 3.1.1: All whitespace should be treated as a single blank space.
+  text = text.replaceAll(_horizontalWhiteSpaceRegex, ' ');
+
+  // Guideline 5: Equivalent Punctuation marks
+  // Replacing punctuations before tokenizing as
+  // SpanScanner deals with one UTF-16 codepoint at at a time
+  // and dealing with surrogate pairs will be problem.
+  // This won't effect either offset, line or column of original text.
+  _equivalentPunctuationMarks.forEach((reg, value) {
+    text = text.replaceAll(reg, value);
+  });
+
+  // Add a space to properly read the last token
+  // without having extra if-conditions.
+  text = text + ' ';
+
+  final _scanner = SpanScanner(text);
+
+  var tokens = <NewToken>[];
+  var tokenID = 0;
+
+  /// Scans through the input text and creates a list of [NewToken]
+  /// Whitespace is ignored but newLine token is stored to deal with list Items.
+  /// Any Leading or standalone punctuation form a separate token
+  /// Example `! !hello!` --> `!`, `!`, `hello!`
+  while (!_scanner.isDone) {
+    var prevState = _scanner.state;
+    var char = _scanner.readChar();
+
+    // Check if the character is alphanumeric,
+    // If it is create a span until you reach space or new Line.
+    // If not create a span for of the single character.
+    // This approach will help us to deal with comment indicators, bullets
+    // and list items.
+    if (_isAlphaNumeric(char)) {
+      while (!_scanner.isDone) {
+        char = _scanner.readChar();
+        if (_isSpace(char) || char == 10) {
+          break;
+        }
+      }
+
+      // Move the scanner back to exclude space or newLine character
+      // from token.
+      _scanner.position--;
+      _addNewToken(_scanner.spanFrom(prevState), tokenID++, tokens);
+
+      prevState = _scanner.state;
+      _scanner.position++;
+
+      if (_isNewLine(char)) {
+        _addNewToken(_scanner.spanFrom(prevState), tokenID++, tokens);
+      }
+    } else if (!_isSpace(char)) {
+      _addNewToken(_scanner.spanFrom(prevState), tokenID++, tokens);
+    }
```

Comment:
If we were to take this a step further we could even handle a lot of the normalization inside `nextToken`. If I wanted to I could easily expand it to handle `_varietalWords` like this:

```dart

final _varietalWords = <String,String>{
  'acknowledgment': 'acknowledgement',
  'analogue': 'analog',
  'analyse': 'analyze',
  'artefact': 'artifact',
  ...
}.map((k, v) => MapEntry( // convert into a Map<RegExp,String>, where the regex ignores case
  RegExp(RegExp.escape(k), caseSensitive: false),
  v,
));

List<Token> tokenize(String originalText) {
  final s = SpanScanner(originalText); // I suggest using the variable name "s" because you'll need to refer to this a lot.

  // Returns the next token
  Token? nextToken() {
    for (final pattern in _varietalWords.keys) {
      if (s.scan(pattern)) {
        final value = _varietalWords[pattern];
        return Token(value, s.lastSpan);
      }
    }

    // Notice how we scan for _varietalWords before we scan for _alphaNumericPattern
    // that's because every instance of _varietalWords would also match _alphaNumericPattern
    // But that's not a problem so long as we scan for _varietalWords first :D

    if (s.scan(_alphaNumericPattern)) {
      return Token(s.lastSpan.text.toLowerCase(), s.lastSpan);
    }

    if (s.scan(_newLinePattern) {
      return Token('\n', s.lastSpan); // ignore the actual text, always normalize to \n
    }

    if (s.scan(_whitespacePattern) {
      return null; // No token here
    }

    for (final pattern in _equivalentPunctuationMarks.keys()) {
      if (s.scan(pattern)) {
        final value = _equivalentPunctuationMarks[pattern];
        return Token(value, s.lastSpan); // always normalize to the equivalent mark
      }
    }

    // TODO: Handle various other cases... for example we need to create tokens for the kinds of punctuations that we don't ignore!
    
    // No pattern matched, we must have a character that is garbage, maybe an emoji or a special character that
    // we are ignore when tokenizing. We'll just read the next char to skip it and return no token.
    s.readChar();
    return null; // No token
  }
```

Reply:
Yes, this approach looks better. I'll adapt to this approach.